File: 2023/04/4.ts
import {
  $,
  add,
  fillArray,
  filter,
  gt,
  int,
  intersect,
  intoSet,
  last,
  length,
  lines,
  map,
  pipe,
  pow,
  range,
  readInput,
  reduce,
  split,
  spy,
  subtract,
  sum,
  tee,
  update
} from '../../common'

$(
  readInput(),
  lines,
  map(pipe(split(/:\s*/), last, split(/\s*\|\s*/), map(pipe(split(/\s+/), map(int), intoSet)), intersect, length)),
  tee(
    pipe(filter(gt(0)), map(pipe(subtract(1), pow(2))), sum),
    pipe(wins =>
      $(
        wins,
        length,
        range,
        reduce(
          (cards, card) =>
            $(
              range(card + 1, wins[card] + card + 1),
              reduce((cs, c) => $(cs, update(c, add(cards[card]))), cards)
            ),
          fillArray(length(wins), 1)
        ),
        sum
      )
    )
  ),
  spy
)

File: 2023/04/4.clj
(ns aoc.2023.04.4 
  (:require [aoc.common :refer [read-input]]
            [clojure.math.numeric-tower :refer [expt]]
            [clojure.set :as set]
            [clojure.string :as str]))

(let [wins (->> (read-input)
                (map #(str/replace % #".*:" ""))
                (map #(str/split % #"\|"))
                (mapv (fn [line]
                        (->> line
                             (map #(re-seq #"\d+" %))
                             (map #(map parse-long %))
                             (map set))))
                (map #(apply set/intersection %))
                (mapv count))

      won-cards-initial (vec (take (count wins) (repeat 1)))
      
      won-cards (->> (range (count wins))
                     (reduce (fn [cards card]
                               (->> (range (inc card) (+ card (get wins card) 1))
                                    (reduce (fn [cs c] (update cs c #(+ % (get cards card)))) cards)))
                             won-cards-initial))]

  (->> wins
       (filter #(> % 0))
       (map #(expt 2 (dec %)))
       (apply +)
       (println "Part 1:"))

  (->> won-cards
       (apply +)
       (println "Part 2:")))